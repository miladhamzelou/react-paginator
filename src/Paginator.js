import React from 'react'
import PropTypes from 'prop-types'
import range from 'lodash.range'
import Button from './Button'
import Infos from './Infos'
import { SIZE_PREFIXES } from './utils'

export default class Paginator extends React.PureComponent {
  static propTypes = {
    breakLabel: PropTypes.string.isRequired,
    className: PropTypes.string,
    condensed: PropTypes.bool.isRequired,
    disabled: PropTypes.bool.isRequired,
    firstAriaLabel: PropTypes.string.isRequired,
    firstLabel: PropTypes.string.isRequired,
    lastAriaLabel: PropTypes.string.isRequired,
    lastLabel: PropTypes.string.isRequired,
    nbRowsPerPage: PropTypes.number.isRequired,
    nextAriaLabel: PropTypes.string.isRequired,
    nextLabel: PropTypes.string.isRequired,
    onPageChange: PropTypes.func.isRequired,
    page: PropTypes.number.isRequired,
    pageRange: PropTypes.number.isRequired,
    previousAriaLabel: PropTypes.string.isRequired,
    previousLabel: PropTypes.string.isRequired,
    showInfos: PropTypes.bool.isRequired,
    size: PropTypes.oneOf(['large', 'medium', 'small']).isRequired,
    total: PropTypes.number.isRequired,
  }

  static defaultProps = {
    breakLabel: '…',
    condensed: false,
    disabled: false,
    firstAriaLabel: 'First',
    firstLabel: '«',
    lastAriaLabel: 'Last',
    lastLabel: '»',
    onPageChange: () => {},
    nextAriaLabel: 'Next',
    nextLabel: '›',
    pageRange: 3,
    previousAriaLabel: 'Previous',
    previousLabel: '‹',
    size: 'medium',
    showInfos: false,
  }

  handlePageChange = (page) => {
    if (page >= 1 && page <= this.getNbPages() && !this.props.disabled) this.props.onPageChange(page)
  }

  getNbPages() {
    const nbPages = Math.ceil(this.props.total / this.props.nbRowsPerPage)
    return nbPages <= 0 ? 1 : nbPages
  }

  getMinRange() {
    return (Math.ceil(this.props.page / this.props.pageRange) - 1) * this.props.pageRange + 1
  }

  getMaxRange() {
    const max = this.getMinRange() + this.props.pageRange - 1
    return max < this.getNbPages() ? max : this.getNbPages()
  }

  renderInfos() {
    if (!this.props.showInfos) return null

    return (
      <Infos
        max={this.props.page * this.props.nbRowsPerPage}
        min={(this.props.page - 1) * this.props.nbRowsPerPage + 1}
        size={this.props.size}
        total={this.props.total}
      />
    )
  }

  renderPages() {
    if (this.props.disabled) return <Button disabled label={this.props.breakLabel} />

    return range(this.getMinRange(), this.getMaxRange() + 1).map(this.renderPage)
  }

  renderPage = (page, index) => (
    <Button
      active={this.props.page === page}
      key={index}
      label={String(page)}
      onPageChange={this.handlePageChange}
      page={page}
    />
  )

  renderPrevious() {
    return this.renderNeighbour({
      ariaLabel: this.props.previousAriaLabel,
      disabled: this.props.page === 1,
      label: this.props.previousLabel,
      page: this.props.page - 1,
    })
  }

  renderNext() {
    return this.renderNeighbour({
      ariaLabel: this.props.nextAriaLabel,
      disabled: this.props.page === this.getNbPages(),
      label: this.props.nextLabel,
      page: this.props.page + 1,
    })
  }

  renderNeighbour({ ariaLabel, disabled, label, page }) {
    return (
      <Button
        ariaLabel={ariaLabel}
        disabled={disabled || this.props.disabled}
        label={label}
        onPageChange={this.handlePageChange}
        page={page}
      />
    )
  }

  renderFirst() {
    return this.renderBoundary({
      ariaLabel: this.props.firstAriaLabel,
      label: this.props.firstLabel,
      page: 1,
    })
  }

  renderLast() {
    return this.renderBoundary({
      ariaLabel: this.props.lastAriaLabel,
      label: this.props.lastLabel,
      page: this.getNbPages(),
    })
  }

  renderBoundary({ ariaLabel, label, page }) {
    return (
      <Button
        ariaLabel={ariaLabel}
        disabled={this.props.disabled || this.props.page === page}
        label={label}
        onPageChange={this.handlePageChange}
        page={page}
      />
    )
  }

  renderPreviousBreak() {
    return this.renderBreak(this.getMinRange() - 1)
  }

  renderNextBreak() {
    return this.renderBreak(this.getMaxRange() + 1)
  }

  renderBreak(page) {
    if (this.props.disabled || page >= this.getNbPages() || page <= 0) return null

    return <Button label={this.props.breakLabel} onPageChange={this.handlePageChange} page={page} />
  }

  render() {
    const sizePrefix = SIZE_PREFIXES[this.props.size]
    const className = `pagination pagination-${sizePrefix}`

    if (this.props.condensed) {
      return (
        <div className={this.props.className}>
          {this.renderInfos()}
          <ul className={className}>
            {this.renderPrevious()}
            {this.renderNext()}
          </ul>
        </div>
      )
    }

    return (
      <div className={this.props.className}>
        {this.renderInfos()}
        <ul className={className}>
          {this.renderFirst()}
          {this.renderPrevious()}
          {this.renderPreviousBreak()}
          {this.renderPages()}
          {this.renderNextBreak()}
          {this.renderNext()}
          {this.renderLast()}
        </ul>
      </div>
    )
  }
}
