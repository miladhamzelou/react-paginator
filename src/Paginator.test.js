import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import Paginator from './'

describe('Paginator', () => {
  describe('labels', () => {
    it('should be possible to specify custom labels', () => {
      const tree = renderer.create(
        <Paginator
          breakLabel="..."
          firstAriaLabel="Premier"
          firstLabel="<<"
          lastAriaLabel="Dernier"
          lastLabel=">>"
          nbRowsPerPage={20}
          nextAriaLabel="Suivant"
          nextLabel=">"
          page={1}
          previousAriaLabel="Précédent"
          previousLabel="<"
          total={300}
        />,
      )
      expect(tree).toMatchSnapshot()
    })
  })

  describe('onPageChange', () => {
    it('should be called when we click on buttons', () => {
      const spy = jest.fn()
      const wrapper = mount(<Paginator nbRowsPerPage={20} onPageChange={spy} page={10} total={300} />)

      const buttons = wrapper.find('li a')

      spy.mockClear()
      buttons.first().simulate('click') // first
      expect(spy).toHaveBeenCalledWith(1)

      spy.mockClear()
      buttons.last().simulate('click') // last
      expect(spy).toHaveBeenCalledWith(15)

      spy.mockClear()
      buttons.at(1).simulate('click') // previous
      expect(spy).toHaveBeenCalledWith(9)

      spy.mockClear()
      buttons.at(buttons.length - 2).simulate('click') // next
      expect(spy).toHaveBeenCalledWith(11)

      spy.mockClear()
      buttons.at(2).simulate('click') // first break
      expect(spy).toHaveBeenCalledWith(9)

      spy.mockClear()
      buttons.at(buttons.length - 3).simulate('click') // last break
      expect(spy).toHaveBeenCalledWith(13)
    })
  })

  describe('condensed', () => {
    it('should display a condensed version of paginator', () => {
      const tree = renderer.create(<Paginator condensed nbRowsPerPage={20} page={10} total={300} />)
      expect(tree).toMatchSnapshot()
    })
  })

  describe('size', () => {
    it('should be medium by default', () => {
      const tree = renderer.create(<Paginator nbRowsPerPage={20} page={10} total={300} />)
      expect(tree).toMatchSnapshot()
    })

    it('should work with small', () => {
      const tree = renderer.create(<Paginator nbRowsPerPage={20} page={10} size="small" total={300} />)
      expect(tree).toMatchSnapshot()
    })

    it('should work with large', () => {
      const tree = renderer.create(<Paginator nbRowsPerPage={20} page={10} size="large" total={300} />)
      expect(tree).toMatchSnapshot()
    })
  })

  describe('showInfos', () => {
    it('should display informations', () => {
      const tree = renderer.create(<Paginator nbRowsPerPage={20} page={10} showInfos size="large" total={300} />)
      expect(tree).toMatchSnapshot()
    })
  })

  describe('className', () => {
    it('should support className', () => {
      const tree = renderer.create(
        <Paginator className="my-classname" nbRowsPerPage={20} page={10} showInfos total={300} />,
      )
      expect(tree).toMatchSnapshot()
    })
  })

  describe('disabled', () => {
    it('should support disabled', () => {
      const spy = jest.fn()
      const wrapper = mount(
        <Paginator disabled nbRowsPerPage={20} onPageChange={spy} page={10} showInfos total={300} />,
      )

      const buttons = wrapper.find('li')
      buttons.forEach((button) => {
        expect(button.hasClass('disabled'))
        button.find('a').simulate('click')
      })

      expect(spy).not.toHaveBeenCalled()
    })
  })
})
