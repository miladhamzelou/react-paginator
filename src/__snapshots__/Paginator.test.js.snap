// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator className should support className 1`] = `
<div
  className="my-classname"
>
  <div
    className="pagination-infos pagination-infos-md"
  >
    181-200 / 300
  </div>
  <ul
    className="pagination pagination-md"
  >
    <li
      className=""
    >
      <a
        aria-label="First"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Previous"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="…"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
      </a>
    </li>
    <li
      className="active"
    >
      <a
        aria-label="10"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          10
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="11"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          11
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="12"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          12
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="…"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Next"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Last"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Paginator condensed should display a condensed version of paginator 1`] = `
<div
  className={undefined}
>
  <ul
    className="pagination pagination-md"
  >
    <li
      className=""
    >
      <a
        aria-label="Previous"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Next"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Paginator labels should be possible to specify custom labels 1`] = `
<div
  className={undefined}
>
  <ul
    className="pagination pagination-md"
  >
    <li
      className="disabled"
    >
      <a
        aria-label="Premier"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          &lt;&lt;
        </span>
      </a>
    </li>
    <li
      className="disabled"
    >
      <a
        aria-label="Précédent"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          &lt;
        </span>
      </a>
    </li>
    <li
      className="active"
    >
      <a
        aria-label="1"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          1
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="2"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          2
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="3"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          3
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="..."
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ...
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Suivant"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          &gt;
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Dernier"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          &gt;&gt;
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Paginator showInfos should display informations 1`] = `
<div
  className={undefined}
>
  <div
    className="pagination-infos pagination-infos-lg"
  >
    181-200 / 300
  </div>
  <ul
    className="pagination pagination-lg"
  >
    <li
      className=""
    >
      <a
        aria-label="First"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Previous"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="…"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
      </a>
    </li>
    <li
      className="active"
    >
      <a
        aria-label="10"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          10
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="11"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          11
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="12"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          12
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="…"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Next"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Last"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Paginator size should be medium by default 1`] = `
<div
  className={undefined}
>
  <ul
    className="pagination pagination-md"
  >
    <li
      className=""
    >
      <a
        aria-label="First"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Previous"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="…"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
      </a>
    </li>
    <li
      className="active"
    >
      <a
        aria-label="10"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          10
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="11"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          11
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="12"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          12
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="…"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Next"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Last"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Paginator size should work with large 1`] = `
<div
  className={undefined}
>
  <ul
    className="pagination pagination-lg"
  >
    <li
      className=""
    >
      <a
        aria-label="First"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Previous"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="…"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
      </a>
    </li>
    <li
      className="active"
    >
      <a
        aria-label="10"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          10
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="11"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          11
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="12"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          12
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="…"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Next"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Last"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Paginator size should work with small 1`] = `
<div
  className={undefined}
>
  <ul
    className="pagination pagination-sm"
  >
    <li
      className=""
    >
      <a
        aria-label="First"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Previous"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="…"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
      </a>
    </li>
    <li
      className="active"
    >
      <a
        aria-label="10"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          10
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="11"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          11
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="12"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          12
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="…"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Next"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
      </a>
    </li>
    <li
      className=""
    >
      <a
        aria-label="Last"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
      </a>
    </li>
  </ul>
</div>
`;
